Add Table Store 0.1.0 release (#531)
apache/flink-web: _posts/2022-05-01-release-table-store-0.1.0.md

---
layout: post
title:  "Apache Flink Table Store 0.1.0 Release Announcement"
subtitle: "Unified streaming and batch store for building dynamic tables on Apache Flink."
date: 2022-05-01T08:00:00.000Z
categories: news
authors:
- Jingsong Lee:
  name: "Jingsong Lee"

---

The Apache Flink community is pleased to announce the preview release of the
[Apache Flink Table Store](https://github.com/apache/flink-table-store) (0.1.0).

Flink Table Store is a unified streaming and batch store for building dynamic tables
on Apache Flink. It uses a full Log-Structured Merge-Tree (LSM) structure for high speed

> and a large amount of data update & query capability.

Reviewer: The description of "unified streaming and batch store" sounds a little bit odd to me, and talking about data structure (LSM-tree) is too detailed. How about changing into something like "Flink Table Store is for building dynamic tables for both stream and batch processing in Flink, supporting high speed data ingestion and timely data query"?

Please check out the full [documentation]({{site.DOCS_BASE_URL}}flink-table-store-docs-release-0.1/) for detailed information and user guides.

Note: Flink Table Store is still in beta status and undergoing rapid development,
we do not recommend that you use it directly in a production environment.

## What is Flink Table Store

Open [Flink official website](https://flink.apache.org/), you can see the following line:
`Apache Flink - Stateful Computations over Data Streams.` Flink focuses on distributed computing,
which brings real-time big data computing. Users need to combine Flink with some kind of external storage.

The message queue will be used in both source & intermediate stages in streaming pipeline, to guarantee the
latency stay within seconds. There will also be a real-time OLAP system receiving processed data in streaming
fashion and serving user’s ad-hoc queries.

Everything works fine as long as users only care about the aggregated results. But when users start to care

> about the intermediate data, they will immediately hit a blocker: Intermediate kafka tables are not queryable.

Reviewer 2: The doc uses `message queue` instead of `Kafka` in the previous part, and we'd better to keep it still as a `message-queue-based tables`. Otherwise, it might make users a bit confused.

> Therefore, users use multiple systems. Writing to a lake store like Apache Hudi, Apache Iceberg while writing to Queue,

Reviewer 2: ```suggestion
Therefore, users use multiple systems. Writing to a lake store like Apache Hudi or Apache Iceberg while writing to Queue,
```

Author: I will delete this sentence

the lake store keeps historical data at a lower cost.

> There are two main issues with doing this:

Reviewer 3: I think there is another critical benefits which is not mentioned.   In the newly introduced flink unified table storage.  We unified both the realtime physical data set and batch offline physical data set into a single physical data set, which means we have to separate the realtime data from the batch offline data because they usually don't have the same table format and we don't have the practical approach to flush those realtime serving data into the batch offline table format in the old architecture. 

In the new architecture, we are trying to share the same data set for both realtime data and batch offline data, and the realtime will be flushed to batch offline data automatically as time advance.  The batch offline data can also be accelerated for the OLAP query. The key benefit in the new architecture is:  we don't need to maintain two different data set for realtime OLAP query and batch offline query,  and people will save lots of ETL processing to transform data between those two kinds of data warehouse.

Author: I mainly think it's better to focus on the problems solved by the current version, which avoids bringing relatively big disappointment to users. We can have some follow-up plans

- High understanding bar for users: It’s also not easy for users to understand all the SQL connectors,
  learn the capabilities and restrictions for each of those. Users may also want to play around with
  streaming & batch unification, but don't really know how, given the connectors are most of the time different
  in batch and streaming use cases.
- Increasing architecture complexity: It’s hard to choose the most suited external systems when the requirements
  include streaming pipelines, offline batch jobs, ad-hoc queries. Multiple systems will increase the operation
  and maintenance complexity. Users at least need to coordinate between the queue system and file system of each

>   table, which is error-prone.

Reviewer 4: It seems that usually a release announcement focus on what we envision for the project and what we offer in this release. Personally I feel the comparison with other projects might not be necessary here.

Reviewer: I share the same feeling, and maybe adding a picture to depict the (better, easier, cleaner - as indicated here) architecture with the flink table store solution could help readers to understand.

Reviewer 5: I also have the same feeling the announcement misses a picture to explain the position and capability of the table store.

Author: > It seems that usually a release announcement focus on what we envision for the project and what we offer in this release. Personally I feel the comparison with other projects might not be necessary here.

You are right, the purpose here is actually to draw out why there is a table store, but it seems too much at the moment, and no other systems should be mentioned.

Author: > I share the same feeling, and maybe adding a picture to depict the (better, easier, cleaner - as indicated here) architecture with the flink table store solution could help readers to understand.

> I also have the same feeling the announcement misses a picture to explain the position and capability of the table store.

We should do this.

The Flink Table Store aims to provide a unified storage abstraction:
- Table Store provides storage of historical data while providing queue abstraction.
- Table Store provides competitive historical storage with lake storage capability, using LSM file structure

>   to store data on DFS, providing real-time updates and queries at a lower cost.

Reviewer 5: It sounds like Table Store only supports storing data on DFS and doesn't support object storage.

Reviewer 3: I think the object storage is also another kind of `DFS`...

Author: I will add cloud storage

Author: > I think the object storage is also another kind of DFS...

Yes, I think too...  we can also add more description

- Table Store coordinates between the queue storage and historical storage, providing hybrid read and write capabilities.
- Table Store is a storage created for Flink, which satisfies all the concepts of Flink SQL and is the most
  suitable storage abstraction for Flink.

## Core Features

Flink Table Store supports the following usage:
- **Streaming Insert**: Write changelog streams, including CDC from the database and streams.
- **Batch Insert**: Write batch data as offline warehouse, including OVERWRITE support.
- **Batch/OLAP Query**: Read the snapshot of the storage, efficient querying of real-time data.

> - **Streaming Query**: Read the storage changes, ensure exactly-once consistency.

Reviewer 2: If we follow the order of `streaming insert` and then `batch insert`, I think we should better to give the order of `streaming query` and then `batch query`.
BTW, do we need to split the features of reading snapshot and real-time data? From my side, they are actually two features.

Author: I reorganized the language

Flink Table Store uses the following technologies to support the above user usages:
- Hybrid Storage: Integrating Apache Kafka to achieve real-time stream computation.
- LSM Structure: For a large amount of data updates and high performance queries.
- Columnar File Format: Use Apache ORC to support efficient querying.

> - Lake Storage: Metadata and data on DFS and Object Store.

Reviewer: I wonder whether it's necessary to expose the implementation details in the release blog post, especially when the table store is still in a preview status and implementations may change in the future.

OTOH, if we think it's still valuable to provide details here, I would suggest to add a "In this preview version" at the beginning of the paragraph.

Reviewer 2: I think adding some details is okay here as this is the 1st version and users would still be curious about the implementations considering not so many guys involved in development.
However, I think the description looks a bit strange. The details of `Hybrid Storage`, `Columnar File Format` and `Lake Storage` are used to explain how we implement it. While the details of `LSM Structure` is used to explain why we need this. Maybe we can use `key-value store` to replace `LSM Structure` and then explain how we implement this.

Author: I'll organize the language and add the architecture picture.

Many thanks for the inspiration of the following systems: [Apache Iceberg](https://iceberg.apache.org/) and [RocksDB](http://rocksdb.org/).

## Getting started

> For a detailed [getting started guide]({{site.DOCS_BASE_URL}}flink-table-store-docs-release-0.1/docs/try-table-store/quick-start/) please check the documentation site.

Reviewer: ```suggestion
Please refer to the [getting started guide]({{site.DOCS_BASE_URL}}flink-table-store-docs-release-0.1/docs/try-table-store/quick-start/) for more details.
```

## What's Next?

The community is currently working on hardening the core logic, stabilizing the storage format and adding the remaining bits for making the Flink Table Store production-ready.

In the upcoming 0.2.0 release you can expect (at-least) the following additional features:

> * Ecology: Support Flink Table Store Reader for Apache Hive Engine

Reviewer 4: Ecology -> Ecosystem?

---
layout: post
title:  "Apache Flink Table Store 0.1.0 Release Announcement"
subtitle: "Unified streaming and batch store for building dynamic tables on Apache Flink."
date: 2022-05-01T08:00:00.000Z
categories: news
authors:
- Jingsong Lee:
  name: "Jingsong Lee"

---

> The Apache Flink Community is pleased to announce the preview release of the

Reviewer 6: 'Community' better be lowercased here.

Ref
![image](https://user-images.githubusercontent.com/55568005/166945854-91d287d0-39d2-4b75-a21b-ac699331ee69.png)

[Apache Flink Table Store](https://github.com/apache/flink-table-store) (0.1.0).

Flink Table Store is a unified streaming and batch store for building dynamic tables
on Apache Flink. It uses a full Log-Structured Merge-Tree (LSM) structure for high speed

> and large amount of data update & query capability.

Reviewer 6: "and a large amount of"

Please check out the full [documentation]({{site.DOCS_BASE_URL}}flink-table-store-docs-release-0.1/) for detailed information and user guides.

Note: Flink Table Store is still in beta status and undergoing rapid development,
we do not recommend that you use it directly in a production environment.

## What is Flink Table Store

Open [Flink official website](https://flink.apache.org/), you can see the following line:
`Apache Flink - Stateful Computations over Data Streams.` Flink focuses on distributed computing,

> which brings real-time big data computing. But pure computation doesn't bring value, users need

Reviewer 6: "doesn't" => "does not" since contractions maybe informal for technical writing.

> which brings real-time big data computing. But pure computation doesn't bring value, users need

Reviewer 6: Nit: "But" => "Nevertheless" or "However"

> which brings real-time big data computing. But pure computation doesn't bring value, users need

Reviewer 6: > But pure computation doesn't bring value

I'm afraid that this expression is not very appropriate? We can highlight the significance of Table Store, but it does not have to sentence pure computing is valueless. What do you think?

Author: I'm not saying this well, I'll change it.
I meant compute + storage to be the complete business.

to combine Flink with some kind of external storage.

For a long time, we found that no external storage can fit Flink's computation model perfectly,

> which brings troubles to users. So Flink Table Store was born, it is a storage built specifically

Reviewer 3: Could you explain a bit more about the real troubles that people encountered in this post ?  I think that will be helpful for others to understand what are the problem that flink table store is trying to resolve ?

Author: I will introduce more from https://cwiki.apache.org/confluence/display/FLINK/FLIP-188%3A+Introduce+Built-in+Dynamic+Table+Storage

> for Flink, for big data real-time update scenario. From now on, Flink is no longer just a computing

Reviewer 6: Nit: "scenarios"

engine.

> Flink Table Store is a unified streaming and batch table format:

Reviewer 3: > Flink Table Store is a unified streaming and batch table format

We are talking about the short-term roadmap, right ?

Author: Yes, but I should change the `table format` here.

- As the storage of Flink, it first provides the capability of Queue.
- On top of the Queue capability, it precipitates historical data to data lakes.
- The data on data lakes can be updated and analyzed in near real-time.

## Core Features

Flink Table Store supports the following usage:

> - **Streaming Insert**: Write changelog streams, including CDC from database and streams.

Reviewer 6: Nit: "from the database"

- **Batch Insert**: Write batch data as offline warehouse, including OVERWRITE support.

> - **Batch/OLAP Query**: Read snapshot of the storage, efficient querying of real-time data.

Reviewer 6: Nit: "the snapshot"

> - **Streaming Query**: Read changes of the storage, ensure exactly-once consistency.

Reviewer 6: Nit: "changes of the storage" may be wordy and can be simplified as "storage changes".

Flink Table Store uses the following technologies to support the above user usages:
- Hybrid Storage: Integrating Apache Kafka to achieve real-time stream computation.

> - LSM Structure: For large amount of data updates and high performance query.

Reviewer 6: Nit: "a large number of" and "high-performance queries"

- Columnar File Format: Use Apache ORC to support efficient querying.
- Lake Storage: Metadata and data on DFS and Object Store.

> Many thanks for the inspiration of the following systems: Apache Iceberg, RocksDB.

Reviewer 6: Nit: "and" instead of ","

Reviewer 3: btw, could please add the official links for those two projects ? Thanks.

## Getting started

For a detailed [getting started guide]({{site.DOCS_BASE_URL}}flink-table-store-docs-release-0.1/docs/try-table-store/quick-start/) please check the documentation site.

## What's Next?

> The community is currently working on hardening the core logic, stabilizing the storage format and adding the remaining bits for making the Flink Table Store production ready.

Reviewer 6: "production-ready"

In the upcoming 0.2.0 release you can expect (at-least) the following additional features:

* Ecology: Support Flink Table Store Reader for Apache Hive Engine
* Core: Support the adjustment of the number of Bucket
* Core: Support for Append Only Data, Table Store is not just limited to update scenarios
* Core: Full Schema Evolution
* Improvements based on feedback from the preview release

> In the medium term you can also expect:

Reviewer 6: Nit: add a comma here, "In the medium term, ..."

* Ecology: Support Flink Table Store Reader for Trino, PrestoDB and Apache Spark

> * Flink Table Store Service to accelerate update and improve query performance

Reviewer 6: "updates"

Please give the preview release a try, share your feedback on the Flink mailing list and contribute to the project!

## Release Resources

The source artifacts and binaries are now available on the updated [Downloads](https://flink.apache.org/downloads.html)
page of the Flink website.

We encourage you to download the release and share your feedback with the community through the [Flink mailing lists](https://flink.apache.org/community.html#mailing-lists)
or [JIRA](https://issues.apache.org/jira/issues/?jql=project%20%3D%20FLINK%20AND%20component%20%3D%20%22Table%20Store%22).

## List of Contributors

> The Apache Flink community would like to thank each and every one of the contributors that have made this release possible:

Reviewer 6: "thank every one"
